# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/events/EventStateManager.cpp
# Commit: 024cf173d87d
# Full Hash: 024cf173d87dbc5d3b581d2b121e3d8295e4f46f
# Author: Dan Robertson <drobertson@mozilla.com>
# Date: 2023-09-26 09:27:49
# Regressor Bug: 1816473
# File Overlap Count: 3
# Description:
#   Bug 1816473 - Do not delay dispatching synthesized mouse events. r=botond,edgar
#   
#   Do not delay dispatching synthesized mouse events for a single tap
#   gesture.
#   
# ==============================================================================

diff -r 37eee8ccc556 -r 024cf173d87d dom/events/EventStateManager.cpp
--- a/dom/events/EventStateManager.cpp	Mon Sep 25 19:08:05 2023 +0000
+++ b/dom/events/EventStateManager.cpp	Mon Sep 25 19:13:07 2023 +0000
@@ -3642,8 +3642,13 @@
       // we should never be capturing when the mouse button is up
       PresShell::ReleaseCapturingContent();
 
-      ClearGlobalActiveContent(this);
       WidgetMouseEvent* mouseUpEvent = aEvent->AsMouseEvent();
+      // If the mouseup event is a synthesized mouse event due to a touch, do
+      // not clear the activation state. Element activation is handled by APZ.
+      if (!mouseUpEvent || mouseUpEvent->mInputSource !=
+                               dom::MouseEvent_Binding::MOZ_SOURCE_TOUCH) {
+        ClearGlobalActiveContent(this);
+      }
       if (mouseUpEvent && EventCausesClickEvents(*mouseUpEvent)) {
         // Make sure to dispatch the click even if there is no frame for
         // the current target element. This is required for Web compatibility.